compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -34,53 +49,240 @@ pub fn expand_deriving_default(
             attributes: attrs,
             is_unsafe: false,
             unify_fieldless_variants: false,
-            combine_substructure: combine_substructure(Box::new(|a, b, c| {
-                default_substructure(a, b, c)
+            combine_substructure: combine_substructure(Box::new(|cx, trait_span, substr| {
+                match substr.fields {
+                    StaticStruct(_, fields) => {
+                        default_struct_substructure(cx, trait_span, substr, fields)
+                    }
+                    StaticEnum(enum_def, _) if cx.sess.features_untracked().derive_default_enum => {
```

Comment:
Features shouldn't generally affect compiler behavior, only report errors.
So the logic should look roughly like this:
```rust
                    StaticEnum(enum_def, _) => {
                        if !cx.sess.features_untracked().derive_default_enum {
                            rustc_session::parse::feature_err(...);
                        }
                        default_enum_substructure(cx, trait_span, enum_def)
                    }
```